core/src/subgraph/instance_manager.rs

```diff
@@ -653,18 +648,39 @@ where
 
     // Process events one after the other, passing in entity operations
     // collected previously to every new event being processed
-    let (mut ctx, mut block_state) = process_triggers(
+    let mut block_state = match process_triggers(
         &logger,
         BlockState::new(
             ctx.inputs.store.clone(),
             std::mem::take(&mut ctx.state.entity_lfu_cache),
         ),
         proof_of_indexing.cheap_clone(),
-        ctx,
+        ctx.subgraph_metrics.clone(),
+        &ctx.state.instance,
         &light_block,
         triggers,
     )
-    .await?;
+    .await
+    {
+        Ok(block_state) => block_state,
+        Err(MappingError::Unknown(e)) => return Err(e.compat_err().into()),
+        Err(MappingError::PossibleReorg(e)) => {
+            info!(ctx.state.logger,
+                    "Possible reorg detected, retrying";
+                    "error" => format!("{:?}", e.to_string()),
+                    "id" => ctx.inputs.deployment_id.to_string(),
+            );
+
+            // In case of a possible reorg, we want this function to do nothing and restart the
+            // block stream so it has a chance to detect the reorg.
+            //
+            // The `ctx` is unchanged at this point, except for having cleared the entity cache.
+            // Losing the cache is a bit annoying but not an issue for correctness.
+            //
+            // See also b21fa73b-6453-4340-99fb-1a78ec62efb1.
+            return Ok((ctx, true));
+        }
```

Comment:
Ok, this answers my previous question. We're bubbling the problem up and retrying the stream as a whole because we may not want to stick to the hash that keeps failing. Makes perfect sense.